integrationtests/server_test.go

```diff
@@ -0,0 +1,201 @@
+package integrationtests
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"io/ioutil"
+	"math/big"
+	mrand "math/rand"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/lucas-clemente/quic-go/h2quic"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gexec"
+)
+
+var _ = Describe("Server tests", func() {
+	mrand.Seed(time.Now().UnixNano())
+
+	var (
+		port    string
+		tmpDir  string
+		session *Session
+		client  *http.Client
+	)
+
+	generateCA := func() (*rsa.PrivateKey, *x509.Certificate) {
+		key, err := rsa.GenerateKey(rand.Reader, 1024)
+		Expect(err).ToNot(HaveOccurred())
+
+		t := time.Now().Add(-time.Minute)
+		templateRoot := &x509.Certificate{
+			SerialNumber: big.NewInt(1),
+			NotBefore:    t,
+			NotAfter:     t.Add(time.Hour),
+			IsCA:         true,
+			BasicConstraintsValid: true,
+		}
+		certDER, err := x509.CreateCertificate(rand.Reader, templateRoot, templateRoot, &key.PublicKey, key)
+		Expect(err).ToNot(HaveOccurred())
+		cert, err := x509.ParseCertificate(certDER)
+		Expect(err).ToNot(HaveOccurred())
+		return key, cert
+	}
+
+	// prepare the file such that it can be by the quic_server
+	// some HTTP headers neeed to be prepended, see https://www.chromium.org/quic/playing-with-quic
+	createDownloadFile := func(filename string, data []byte) {
+		dataDir := filepath.Join(tmpDir, "quic.clemente.io")
+		err := os.Mkdir(dataDir, 0777)
+		Expect(err).ToNot(HaveOccurred())
+		f, err := os.Create(filepath.Join(dataDir, filename))
+		Expect(err).ToNot(HaveOccurred())
+		defer f.Close()
+		_, err = f.Write([]byte("HTTP/1.1 200 OK\n"))
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write([]byte("Content-Type: text/html\n"))
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write([]byte("X-Original-Url: https://quic.clemente.io:" + port + "/" + filename + "\n"))
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write([]byte("Content-Length: " + strconv.Itoa(len(data)) + "\n\n"))
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write(data)
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	// download files must be create *before* the quic_server is started
+	// the quic_server reads its data dir on startup, and only serves those files that were already present then
+	startServer := func() {
+		go func() {
+			defer GinkgoRecover()
+			var err error
+			command := exec.Command(
+				serverPath,
+				"--quic_response_cache_dir="+filepath.Join(tmpDir, "quic.clemente.io"),
+				"--key_file="+filepath.Join(tmpDir, "key.pkcs8"),
+				"--certificate_file="+filepath.Join(tmpDir, "cert.pem"),
+				"--port="+port,
+			)
+			session, err = Start(command, GinkgoWriter, GinkgoWriter)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+	}
+
+	stopServer := func() {
+		session.Kill()
+	}
+
+	BeforeEach(func() {
+		port = strconv.Itoa(20000 + int(mrand.Int31n(10000)))
```

Comment:
Unfortunately, this `quic_server` doesn't support that. If I omit the `--port`, it listens on 6121 by default.